applications/OptimizationApplication/custom_python/add_custom_utilities_to_python.cpp

```diff
@@ -14,22 +14,89 @@
+        .def(py::init<ModelPart&>(), py::arg("model_part"), py::doc("Creates a new container data object with model_part."))
+        .def(py::init<const container_type&>(), py::arg("other_container_data_to_copy_from"), py::doc("Creates a new same type container data object by copying data from other_container_data_to_copy_from."))
+        .def(py::init<const typename container_type::BaseType&>(), py::arg("other_container_data_to_copy_from"), py::doc("Creates a new destination type container data object by copying data from compatible other_container_data_to_copy_from."))
+        .def("AssignDataToContainerVariable", &container_type::template AssignDataToContainerVariable<double>, py::arg("scalar_variable"))
+        .def("AssignDataToContainerVariable", &container_type::template AssignDataToContainerVariable<array_1d<double, 3>>, py::arg("Array3_variable"))
+        .def("ReadDataFromContainerVariable", &container_type::template ReadDataFromContainerVariable<double>, py::arg("scalar_variable"))
+        .def("ReadDataFromContainerVariable", &container_type::template ReadDataFromContainerVariable<array_1d<double, 3>>, py::arg("Array3_variable"))
+        .def("SetDataForContainerVariable", &container_type::template SetDataForContainerVariable<double>, py::arg("scalar_variable"), py::arg("scalar_value"))
+        .def("SetDataForContainerVariable", &container_type::template SetDataForContainerVariable<array_1d<double, 3>>, py::arg("Array3_variable"), py::arg("Array3_value"))
+        .def("SetDataForContainerVariableToZero", &container_type::template SetDataForContainerVariableToZero<double>, py::arg("scalar_variable"))
+        .def("SetDataForContainerVariableToZero", &container_type::template SetDataForContainerVariableToZero<array_1d<double, 3>>, py::arg("Array3_variable"))
+        .def("Clone", &container_type::Clone)
+        .def("CloneWithDataInitializedToZero", &container_type::CloneWithDataInitializedToZero)
+        .def(py::self +  py::self)
+        .def(py::self += py::self)
+        .def(py::self +  float())
+        .def(py::self += float())
+        .def(py::self -  py::self)
+        .def(py::self -= py::self)
+        .def(py::self -  float())
+        .def(py::self -= float())
+        .def(py::self *  py::self)
+        .def(py::self *= py::self)
+        .def(py::self *  float())
+        .def(py::self *= float())
+        .def(py::self /  py::self)
+        .def(py::self /= py::self)
+        .def(py::self /  float())
+        .def(py::self /= float())
+        .def("__pow__", py::overload_cast<const container_type&>(&container_type::operator^, py::const_))
+        .def("__ipow__", py::overload_cast<const container_type&>(&container_type::operator^=))
+        .def("__pow__", py::overload_cast<const double>(&container_type::operator^, py::const_))
+        .def("__ipow__", py::overload_cast<const double>(&container_type::operator^=))
+        .def("__neg__", [](container_type& rSelf) { return rSelf.operator*(-1.0); })
```

Comment:
I don't like exposing these primitive operations on enormous arrays to python. It'll have terrible performance if you'll try chaining operations because there's no expression template system in python and you'll need to pass through the entire array for each operation.

I'd rather go for some system that represents the sequence of operations, and can evaluate them all at once (like expression templates in C++).

Reply:
done :)